### ISSUE TITLE
Canvas width/height discusion

### ISSUE THREAD

**Issue by jordi-z80** [kniEngine/kni#1729]
KNI 3.12.9001 (nuget) , Windows10 , BlazorGL

Hi,

I don't know much of the internals of KNI BlazorGL , so take this suggestion with a grain of salt - I may not be aware of the implications of the changes or even if it's possible.

However, I think modifying the canvas width and height on resize is a wrong approach in HTML, and would be a lot better to let the user set that data on the canvas definition, and default it to 100%.

In example, setting the canvas like this:

`<canvas id="theCanvas" style="touch-action:none; width: 100%; height: 100%;"></canvas>`

and then modifyiing the following  in BlazorGameWindow.cs

```
                //_canvas.Width = _window.InnerWidth;
                //_canvas.Height = _window.InnerHeight;
				
                int newWidth  = _window.InnerWidth;
                int newHeight = _window.InnerHeight;
```

You get the same result (the browser will always do its best to make the canvas 100% size), but you allow the user to use, in example, a width of 50%, or add a margin, or whatever they want without having to enclose the game into an iframe.

The only problem with this approach is that on resize the vertical starting position of the KNI output is offseted, but I think this is probably because of how GL works and may be just modifying an offset that you're already applying (I haven't been able to find it though, so I just don't know if it's possible)

Thanks for such a wonderful library!

EDIT: Actually, some of the problems seem to be with the canvasHolder too. My solution may not be good enough, but I really believe that sizes shouldn't be set by anything other than the browser so that users can leverage HTML/CSS to be able to redesign the Index.razor

**Comment by jordi-z80** [comment]
Sorry for being a pain in the ass lately :)

I've done a small HTML as a sample as how the html/css could be set up so that it gives a lot of flexibility for the user and still be able to use the canvas.

The sample tries to maintain 16/9 , avoid being clamped outside the window, and be able to be placed wherever the user wants (I've added a 80% width size as a sample, but it can be bigger, smaller, or even remove it).

I think without the "backend" enforcing the canvas' width and height (and supposing that offset problem can be fixed) this would work, and you'd still receive the messages from the canvas changing the size, so that the backbuffer can be correctly set up.

https://monstre.boo/files/sample.html

**Comment by nkast** [comment]
My understanding is that Canvas.Width is not an HTMLElement property. It updates the backbuffer (aka swapchain) resolution.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas#sizing_the_canvas_using_css_versus_html
"The displayed size of the canvas can be changed using CSS, but if you do this the image is scaled during rendering to fit the styled size, which can make the final graphics rendering end up being distorted.
It is better to specify your canvas dimensions by setting the width and height attributes directly on the <canvas> elements, either directly in the HTML or by using JavaScript."

By design the canvas size is the same as the client window size. change that would cause a lot of issues.
Mouse/Touch input for example would be offset, and other unforeseeable situations. 
I wrote something about it here https://github.com/kniEngine/kni/issues/261

**Comment by jordi-z80** [comment]
I see, I think I understand the problem now, although I think that what I'm trying to do can still be achieved.

Do you have any document on how to create a BlazorGL + KNI project without using the nugets? I'm trying to build a project that is using my compiled branch of KNI with minor modifications, but I've failed to get it working on BlazorGL.

**Comment by jordi-z80** [comment]
I'm closing this because I figured out how to use KNI with custom compiled versions... and it's actually the way it should be. I had an error when using them, but actually, it was not a KNI problem when increasing the versions, it was a weird bug in VS2022 when using it along Microsoft.AspNetCore.Components.WebAssembly > 8.0.3 IIRC. 

Once I updated VS, I had no problem with KNI custom compiled versions.

About the other part of the problem (resizing the canvas) after some thought I think you're right, and after all, most games are enclosed into an iframe, so it makes little to no sense to try to mess with it :)

Thanks again!
